### PR TITLE
attempt to commonize Kubeflow jobs Multikueue support methods

### PR DESCRIPTION
**Pull request by mszadkow**
/kind cleanup
#### What this PR does / why we need it:
Attempt to unify code so it's not repeated for each kubeflow job multikueue adapter
#### Which issue(s) this PR fixes:

Relates #2552 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/jobframework/interface.go [hunk 1]
```diff
@@ -172,6 +172,11 @@ type MultiKueueAdapter interface {
 	GVK() schema.GroupVersionKind
 }
 
+type UpdateRemoteJob interface {
+	UpdateRemoteJobStatus(localJob, remoteJob interface{})
```
Comment (trasc) [inline pkg/controller/jobframework/interface.go:176]: we can make them client.Object

File: pkg/controller/jobs/kubeflow/common/common.go
Comment (trasc) [inline pkg/controller/jobs/kubeflow/common/common.go:1]: why not put this in the kubeflow package?

File: pkg/controller/jobframework/interface.go [hunk 2]
```diff
@@ -172,6 +172,11 @@ type MultiKueueAdapter interface {
 	GVK() schema.GroupVersionKind
 }
 
+type UpdateRemoteJob interface {
```
Comment (trasc) [inline pkg/controller/jobframework/interface.go:175]: it's place should be somewhere in pkg/controllers/jobs/kubeflow

File: pkg/controller/jobs/kubeflow/common/common.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	clientutil "sigs.k8s.io/kueue/pkg/util/client"
+)
+
+type objAsPtr[T any] interface {
+	metav1.Object
+	client.Object
+	*T
+}
+
+func SyncJob[PtrT objAsPtr[T], T any](
```
Comment (trasc) [inline pkg/controller/jobs/kubeflow/common/common.go:25]: why not put this in a generic struct that implements multikueue adapter and has two members holding the specialized "Update..." functions 

then the job will just need to define the two functions and call a `MakeAdaptor(f1, f2)`.

File: pkg/controller/jobframework/interface.go [hunk 4]
```diff
@@ -172,6 +172,11 @@ type MultiKueueAdapter interface {
...
+	UpdateRemoteJobStatus(localJob, remoteJob interface{}) 
+	UpdateRemoteJobSpec(localJob, remoteJob interface{})
```
Comment (tenzen-y) [inline pkg/controller/jobframework/interface.go:177]: We should not expose these interfaces just for kubeflow jobs.

**Review by tenzen-y — COMMENTED** [review]
That doesn't seem to be my expectations.
Could we commonize kubeflow adapters like kubeflow job reconcilers?

https://github.com/kubernetes-sigs/kueue/tree/main/pkg/controller/jobs/kubeflow/kubeflowjob

File: test/integration/multikueue/multikueue_test.go [hunk 5]
```diff
@@ -897,7 +897,7 @@ var _ = ginkgo.Describe("Multikueue", ginkgo.Ordered, ginkgo.ContinueOnFailure,
 							Succeeded: 1,
 						},
 					}))
-			}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
```
Comment (trasc) [inline test/integration/multikueue/multikueue_test.go:900]: do we still need this? or it's just some debug attempt leftover?

File: pkg/controller/jobs/kubeflow/common/common.go
Comment (trasc) [reply pkg/controller/jobs/kubeflow/common/common.go:1]: I meant `pkg/controller/jobs/kubeflow/kubeflowjob`

Comment (mszadkow) [reply pkg/controller/jobs/kubeflow/common/common.go:1]: Yeah, that makes sense

**Comment by mszadkow** [comment]
@alculquicondor you may want to have a look here too ;)

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+	// add the prebuilt workload
+	labels := remoteJob.GetLabels()
+	if remoteJob.GetLabels() == nil {
+		labels = make(map[string]string, 2)
+	}
+	labels[constants.PrebuiltWorkloadLabel] = workloadName
+	labels[kueuealpha.MultiKueueOriginLabel] = origin
+	remoteJob.SetLabels(labels)
+
```
Comment (alculquicondor) [inline pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go:121]: I think this code should even be part of the MK workload controller, but that can be left for a follow up.

**Review by alculquicondor — COMMENTED** [review]
looks good overall :)

**Comment by alculquicondor** [comment]
/release-note-edit

```release-note
NONE
```

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+func (a adapter[PtrT, T]) GVK() schema.GroupVersionKind {
+	return a.gwk
+}
+
+func (a adapter[PtrT, T]) KeepAdmissionCheckPending() bool {
+	return false
+}
+
+func (a adapter[PtrT, T]) IsJobManagedByKueue(_ context.Context, _ client.Client, _ types.NamespacedName) (bool, string, error) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go:80]:
```suggestion
func (a adapter[PtrT, T]) IsJobManagedByKueue(context.Context, client.Client, types.NamespacedName) (bool, string, error) {
```

**Review by tenzen-y — COMMENTED** [review]
Overall, looks good to me.
This is a comment for a followup.

Could you work on consolidating / commonizing the kubeflow MK integration tests (TFJob/PaddleJob/XGBoostJob/PyTorchJob) https://github.com/kubernetes-sigs/kueue/blob/5b9ff8e74f19296174cf20d69b485f20f8d7c7b0/test/integration/multikueue/multikueue_test.go#L781?

I'm wondering if we can reduce the kubeflow MK integration tests like kubeflowjob reconcilers.

**Comment by mszadkow** [comment]
> Overall, looks good to me. This is a comment for a followup.
> 
> Could you work on consolidating / commonizing the kubeflow MK integration tests (TFJob/PaddleJob/XGBoostJob/PyTorchJob)
> 
> https://github.com/kubernetes-sigs/kueue/blob/5b9ff8e74f19296174cf20d69b485f20f8d7c7b0/test/integration/multikueue/multikueue_test.go#L781
> 
> ?
> I'm wondering if we can reduce the kubeflow MK integration tests like kubeflowjob reconcilers.

I actually work on this one while doing MPIJob multikueue adapter, but I can add it here.

**Comment by tenzen-y** [comment]
> > Overall, looks good to me. This is a comment for a followup.
> > Could you work on consolidating / commonizing the kubeflow MK integration tests (TFJob/PaddleJob/XGBoostJob/PyTorchJob)
> > https://github.com/kubernetes-sigs/kueue/blob/5b9ff8e74f19296174cf20d69b485f20f8d7c7b0/test/integration/multikueue/multikueue_test.go#L781
> > 
> > ?
> > I'm wondering if we can reduce the kubeflow MK integration tests like kubeflowjob reconcilers.
> 
> I actually work on this one while doing MPIJob multikueue adapter, but I can add it here.

It seems that this PR will be finalized soon. So, it may be helpful to work on test refactoring at another PR so that we do not block merging this PR.

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go (hunk 6 above)
Comment (mszadkow) [reply pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_multikueue_adapter.go:121]: ok, noted

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**